### PR TITLE
Fix a bug on http

### PR DESCRIPTION
**Pull request by lunny**
As title

### CODE REVIEW

**Review by zeripath — APPROVED** [review]

File: routers/repo/http.go [hunk 1]
```diff
@@ -437,8 +438,17 @@ func (h *serviceHandler) setHeaderCacheForever() {
...
-	reqFile := path.Join(h.dir, file)
+	reqFile := filepath.Join(h.dir, file)
+	rel, err := filepath.Rel(h.dir, reqFile)
+	if err != nil {
+		http.Error(h.w, "Detect if path available failed", 500)
+		return
+	} else if rel == ".." || strings.HasPrefix(filepath.ToSlash(rel), "../") {
+		http.NotFound(h.w, h.r)
+		return
+	}
```
Comment (zeripath) [inline routers/repo/http.go]: I'm not sure that this is strictly necessary anymore after the changes in GetTextFile but it's potentially good practice

File: routers/repo/http.go [hunk 2]
```diff
@@ -437,8 +438,17 @@ func (h *serviceHandler) setHeaderCacheForever() {
 	h.w.Header().Set("Cache-Control", "public, max-age=31536000")
 }
 
+// sendFile send the provided content from a file
 func (h *serviceHandler) sendFile(contentType, file string) {
-	reqFile := path.Join(h.dir, file)
+	reqFile := filepath.Join(h.dir, file)
+	rel, err := filepath.Rel(h.dir, reqFile)
+	if err != nil {
+		http.Error(h.w, "Detect if path available failed", 500)
```
Comment (zeripath) [inline routers/repo/http.go]: I don't really like the phrasing here.

```suggestion
		http.Error(h.w, "Invalid path requested", 500)
```

File: routers/repo/http.go [hunk 3]
```diff
@@ -437,8 +438,17 @@ func (h *serviceHandler) setHeaderCacheForever() {
...
-	reqFile := path.Join(h.dir, file)
+	reqFile := filepath.Join(h.dir, file)
+	rel, err := filepath.Rel(h.dir, reqFile)
+	if err != nil {
+		http.Error(h.w, "Invalid path requested", 500)
+		return
+	} else if rel == ".." || strings.HasPrefix(filepath.ToSlash(rel), "../") {
+		http.NotFound(h.w, h.r)
+		return
+	}
```
Comment (zeripath) [inline routers/repo/http.go]: Not certain this is strictly necessary anymore after the changes in GetTextFile.

**Comment by kdumontnu** [comment]
I hate to be a pain, but...  can you please include full information for this (and other new PRs)? Specifically, what is the bug?
I monitor active PRs very closely to make sure I understand upstream issues and new features, but many say simply "fix bug" ,"as title", "new route", etc. 

Improving this communication will also help reduce the number of queries in discord answered by "that bug was resolved in X".

**Comment by zeripath** [comment]
So after trying this again I think we were actually OK without this:

```
func ServeFile(w ResponseWriter, r *Request, name string) {
	if containsDotDot(r.URL.Path) {
		// Too many programs use r.URL.Path to construct the argument to
		// serveFile. Reject the request under the assumption that happened
		// here and ".." may not be wanted.
		// Note that name might not contain "..", for example if code (still
		// incorrectly) used filepath.Join(myDir, r.URL.Path).
		Error(w, "invalid URL path", StatusBadRequest)
		return
	}
	dir, file := filepath.Split(name)
	serveFile(w, r, Dir(dir), file, false)
}
```

File: routers/repo/http.go [hunk 4]
```diff
@@ -627,7 +637,7 @@ func GetTextFile(p string) func(*context.Context) {
 			h.setHeaderNoCache()
 			file := ctx.Params("file")
 			if file != "" {
-				h.sendFile("text/plain", "objects/info/"+file)
+				h.sendFile("text/plain", "objects/info/"+path.Clean("/" + file)[1:])
```
Comment (delvh) [inline routers/repo/http.go]:
```suggestion
				h.sendFile("text/plain", "objects/info/"+path.Clean(file))
```

**Comment by wxiaoguang** [comment]
> So after trying this again I think we were actually OK without this:
> 
> ```go
> func ServeFile(w ResponseWriter, r *Request, name string) {
> 	if containsDotDot(r.URL.Path) {
> 		// Too many programs use r.URL.Path to construct the argument to
> 		// serveFile. Reject the request under the assumption that happened
> 		// here and ".." may not be wanted.
> 		// Note that name might not contain "..", for example if code (still
> 		// incorrectly) used filepath.Join(myDir, r.URL.Path).
> 		Error(w, "invalid URL path", StatusBadRequest)
> 		return
> 	}
> 	dir, file := filepath.Split(name)
> 	serveFile(w, r, Dir(dir), file, false)
> }
> ```

@zeripath  We still need a good patch. Because `sendFile` send some headers if the file exists, which may leak file information on server.

File: routers/repo/http.go (hunk 3 above)
Comment (wxiaoguang) [reply routers/repo/http.go]: We have good practices like this:

```
func (l *LocalStorage) buildLocalPath(p string) string {
	return filepath.Join(l.dir, path.Clean("/" + strings.ReplaceAll(p, "\\", "/"))[1:])
}
```

File: routers/repo/http.go (hunk 4 above)
Comment (wxiaoguang) [reply routers/repo/http.go]:
```suggestion
				h.sendFile("text/plain", "objects/info/"+file)
```

Leave it as before, let `sendFile` handle all cases correctly.

File: routers/repo/http.go (hunk 3 above)
Comment (wxiaoguang) [inline routers/repo/http.go]:
```suggestion
	reqFile := filepath.Join(h.dir, path.Clean("/" + strings.ReplaceAll(p, "\\", "/"))[1:])
```

**Comment by wxiaoguang** [comment]
Hmm, it's outdated, already fixed by #16117 , can be closed

**Review by delvh — APPROVED** [review]

**Pull request closed without merging**